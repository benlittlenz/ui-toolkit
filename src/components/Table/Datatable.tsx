import React from 'react';

import { Body } from './Body';
import { Column } from './Column';
import { defaultProps } from './defaultProps';
import { Header, HeaderRow } from './Header';
import { Pagination } from './Pagination';
import { tableReducer } from './reducers/reducer';
import { Action, SingleRowAction } from './reducers/types';
import { TableRow } from './Row';
import { Table } from './Table';
import { TableProps, TableState } from './types';
import { getNumberOfPages } from './utils';

export function DataTable<T>(props: TableProps<T>): JSX.Element {
  const {
    data = defaultProps.data,
    columns = defaultProps.columns,
    keyField = defaultProps.keyField,
    selectableRowsSingle = defaultProps.selectableRowsSingle,
    pagination = defaultProps.pagination,
    // paginationTotalRows = defaultProps.paginationTotalRows,
    paginationDefaultPage = defaultProps.paginationDefaultPage,
    // paginationResetDefaultPage = defaultProps.paginationResetDefaultPage,
    paginationPerPage = defaultProps.paginationPerPage,
    // paginationRowsPerPageOptions = defaultProps.paginationRowsPerPageOptions,
    // paginationIconLastPage = defaultProps.paginationIconLastPage,
    // paginationIconFirstPage = defaultProps.paginationIconFirstPage,
    // paginationIconNext = defaultProps.paginationIconNext,
    // paginationIconPrevious = defaultProps.paginationIconPrevious,
    // paginationComponent = defaultProps.paginationComponent,
    // paginationComponentOptions = defaultProps.paginationComponentOptions,
  } = props;

  const [{ selectedRows, currentPage, rowsPerPage }, dispatch] = React.useReducer<
    React.Reducer<TableState<T>, Action<T>>
  >(tableReducer, {
    // Checkbox
    allSelected: false,
    selectedCount: 0,
    selectedRows: [],
    toggleOnSelectedRowsChange: false,
    // Pagination
    currentPage: paginationDefaultPage,
    rowsPerPage: paginationPerPage,
  });

  const sortedData = React.useMemo(() => {
    return [...data].sort();
  }, [data]);

  const tableRows = React.useMemo(() => {
    if (pagination) {
      const lastIndex = currentPage * rowsPerPage;
      const firstIndex = lastIndex - rowsPerPage;

      return sortedData.slice(firstIndex, lastIndex);
    }
    return sortedData;
  }, [sortedData, pagination, currentPage, rowsPerPage]);

  const handleChangeRowsPerPage = React.useCallback(
    (newRowsPerPage: number) => {
      console.log('newRowsPerPage', newRowsPerPage);
      const rowCount = tableRows.length;
      const updatedPage = getNumberOfPages(rowCount, newRowsPerPage);
      const recalculatedPage = Math.min(currentPage, updatedPage);

      dispatch({
        type: 'CHANGE_ROWS_PER_PAGE',
        page: recalculatedPage,
        rowsPerPage: newRowsPerPage,
      });
    },
    [currentPage, tableRows.length]
  );

  const handlePageChange = React.useCallback((page: number) => {
    dispatch({
      type: 'CHANGE_PAGE',
      page,
    });
  }, []);

  const handleSelectedRow = React.useCallback((action: SingleRowAction<T>) => {
    dispatch(action);
  }, []);

  return (
    <div>
      <Table role="table">
        <Header role="row-group">
          <HeaderRow role="row">
            {columns.map((column) => (
              <Column key={column.id} column={column} />
            ))}
          </HeaderRow>
        </Header>
        <Body role="row-group">
          {tableRows.map((row, i) => {
            console.log('ROW', row);
            console.log('selectedRows', selectedRows);
            const isRowSelected = selectedRows.some((r) => r === row);

            return (
              <TableRow
                key={i}
                id={i}
                rowIndex={i}
                columns={columns}
                row={row}
                keyField={keyField}
                selected={isRowSelected}
                selectableRowsSingle={selectableRowsSingle}
                onSelectedRow={handleSelectedRow}
              />
            );
          })}
        </Body>
      </Table>
      {data.length > 0 && pagination && (
        <Pagination
          currentPage={currentPage}
          rowsPerPage={rowsPerPage}
          rowCount={sortedData.length}
          onPageChange={handlePageChange}
          onChangeRowsPerPage={handleChangeRowsPerPage}
        />
      )}
    </div>
  );
}
