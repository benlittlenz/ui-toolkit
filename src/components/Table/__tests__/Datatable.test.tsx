import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { DataTable } from '../Datatable';

const dataMock = () => {
  return {
    columns: [{ name: 'Name', selector: (row: { name: string }) => row.name }],
    data: [
      {
        id: 1,
        name: 'Apple',
      },
      {
        id: 2,
        name: 'Zuchinni',
      },
    ],
  };
};

test('it should render an empty table', () => {
  const { container } = render(<DataTable data={[]} columns={[]} />);

  expect(container.firstChild).toMatchSnapshot();
});

test('it should render correctly with the selector function', () => {
  type TestData = {
    id: number;
    name: string;
  };
  const data: TestData[] = [{ id: 1, name: 'Joe Doe ' }];
  const columns = [{ name: 'Name', selector: (row: TestData) => row.name }];

  const { container } = render(<DataTable data={data} columns={columns} />);
  expect(container.firstChild).toMatchSnapshot();
});

describe('onSelectedRowsChange method', () => {
  test('it should call onSelectedRowsChange with correct values when select all has been selected', () => {
    const mock = dataMock();
    const updatedMock = jest.fn();

    render(
      <DataTable
        data={mock.data}
        columns={mock.columns}
        selectableRows
        onSelectedRowsChange={updatedMock}
      />
    );

    userEvent.click(screen.getByRole('checkbox', { name: 'select-all-rows' }));

    expect(updatedMock).toBeCalledWith({
      allSelected: true,
      selectedCount: 2,
      selectedRows: mock.data,
    });
  });
  test('it should call onSelectedRowsChange with correct values when all rows have been selected', () => {
    const mock = dataMock();
    const onSelectedRowsChange = jest.fn();

    render(
      <DataTable
        data={mock.data}
        columns={mock.columns}
        selectableRows
        onSelectedRowsChange={onSelectedRowsChange}
      />
    );

    userEvent.click(screen.getByRole('checkbox', { name: 'select-row-1' }));

    expect(onSelectedRowsChange).toBeCalledWith({
      allSelected: false,
      selectedCount: 1,
      selectedRows: [
        {
          id: 1,
          name: 'Apple',
        },
      ],
    });
    userEvent.click(screen.getByRole('checkbox', { name: 'select-row-2' }));

    expect(onSelectedRowsChange).toBeCalledWith({
      allSelected: true,
      selectedCount: 2,
      selectedRows: [
        {
          id: 2,
          name: 'Zuchinni',
        },
        {
          id: 1,
          name: 'Apple',
        },
      ],
    });
  });
});

test('it should update state when props change', () => {
  const mock = dataMock();
  const { container, rerender } = render(<DataTable data={mock.data} columns={mock.columns} />);

  rerender(<DataTable data={[{ id: 3, name: 'Something else' }]} columns={mock.columns} />);

  expect(container.firstChild).toMatchSnapshot();
});

describe('Datatable columns', () => {
  test('it renders correctly with columns/data', () => {
    const mock = dataMock();
    const { container } = render(<DataTable data={mock.data} columns={mock.columns} />);

    expect(container).toMatchSnapshot();
  });
});
