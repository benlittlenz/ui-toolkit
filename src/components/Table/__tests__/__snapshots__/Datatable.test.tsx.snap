// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Datatable columns it renders correctly with columns/data 1`] = `
<div>
  <div>
    <div
      class="sc-hBUSln isUOUn"
      role="table"
    >
      <div
        class="sc-gKclnd cFsFpL"
        role="row-group"
      >
        <div
          class="sc-iCfMLu hSbAuc"
          role="row"
        >
          <div
            class="sc-jRQBWg bwoBBf"
          >
            Name
          </div>
        </div>
      </div>
      <div
        class="sc-bdvvtL hjGchb"
        role="row-group"
      >
        <div
          class="sc-ksdxgE inUDjc"
          id="row-0"
          role="row"
        >
          <div
            class="sc-gsDKAQ sc-dkPtRN sc-hKwDye iigSTN juQwsB flAfZy"
            id="cell-undefined"
            role="grid-cell"
          >
            <div>
              Apple
            </div>
          </div>
        </div>
        <div
          class="sc-ksdxgE inUDjc"
          id="row-1"
          role="row"
        >
          <div
            class="sc-gsDKAQ sc-dkPtRN sc-hKwDye iigSTN juQwsB flAfZy"
            id="cell-undefined"
            role="grid-cell"
          >
            <div>
              Zuchinni
            </div>
          </div>
        </div>
      </div>
    </div>
    <nav
      class="sc-jrQzAO dZcCqi"
    >
      <div
        class="sc-pVTFL gUaKsY"
      >
        <select
          class="sc-furwcr biuCpP"
        >
          <option
            selected=""
            value="5"
          >
            5
          </option>
          <option
            value="10"
          >
            10
          </option>
          <option
            value="25"
          >
            25
          </option>
        </select>
        <svg
          aria-hidden="true"
          fill="currentColor"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M5.293 7.293a1 1 0 011.414 0L10 10.586l3.293-3.293a1 1 0 111.414 1.414l-4 4a1 1 0 01-1.414 0l-4-4a1 1 0 010-1.414z"
            fill-rule="evenodd"
          />
        </svg>
      </div>
      <span
        class="sc-crHmcD sc-egiyK dOoxwS dIFsGL"
      >
        1 of 1
      </span>
      <div
        class="sc-kDTinF jFDtol"
      >
        <button
          aria-disabled="true"
          aria-label="First Page"
          class="sc-iqseJM jtEzsa"
          disabled=""
          id="pagination-first-page"
          type="button"
        >
          <svg
            aria-hidden="true"
            fill="currentColor"
            style="height: 12px; width: 12px;"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M15.707 15.707a1 1 0 01-1.414 0l-5-5a1 1 0 010-1.414l5-5a1 1 0 111.414 1.414L11.414 10l4.293 4.293a1 1 0 010 1.414zm-6 0a1 1 0 01-1.414 0l-5-5a1 1 0 010-1.414l5-5a1 1 0 011.414 1.414L5.414 10l4.293 4.293a1 1 0 010 1.414z"
              fill-rule="evenodd"
            />
          </svg>
        </button>
        <button
          aria-disabled="true"
          aria-label="Previous Page"
          class="sc-iqseJM jtEzsa"
          disabled=""
          id="pagination-previous-page"
          type="button"
        >
          <svg
            aria-hidden="true"
            fill="currentColor"
            style="height: 12px; width: 12px;"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M12.707 5.293a1 1 0 010 1.414L9.414 10l3.293 3.293a1 1 0 01-1.414 1.414l-4-4a1 1 0 010-1.414l4-4a1 1 0 011.414 0z"
              fill-rule="evenodd"
            />
          </svg>
        </button>
        <button
          aria-disabled="true"
          aria-label="Next Page"
          class="sc-iqseJM jtEzsa"
          disabled=""
          id="pagination-next-page"
          type="button"
        >
          <svg
            aria-hidden="true"
            fill="currentColor"
            style="height: 12px; width: 12px;"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M7.293 14.707a1 1 0 010-1.414L10.586 10 7.293 6.707a1 1 0 011.414-1.414l4 4a1 1 0 010 1.414l-4 4a1 1 0 01-1.414 0z"
              fill-rule="evenodd"
            />
          </svg>
        </button>
        <button
          aria-disabled="true"
          aria-label="Last Page"
          class="sc-iqseJM jtEzsa"
          disabled=""
          id="pagination-last-page"
          type="button"
        >
          <svg
            aria-hidden="true"
            fill="currentColor"
            style="height: 12px; width: 12px;"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M10.293 15.707a1 1 0 010-1.414L14.586 10l-4.293-4.293a1 1 0 111.414-1.414l5 5a1 1 0 010 1.414l-5 5a1 1 0 01-1.414 0z"
              fill-rule="evenodd"
            />
            <path
              clip-rule="evenodd"
              d="M4.293 15.707a1 1 0 010-1.414L8.586 10 4.293 5.707a1 1 0 011.414-1.414l5 5a1 1 0 010 1.414l-5 5a1 1 0 01-1.414 0z"
              fill-rule="evenodd"
            />
          </svg>
        </button>
      </div>
    </nav>
  </div>
</div>
`;

exports[`it should render an empty table 1`] = `
<div>
  <div
    class="sc-hBUSln isUOUn"
    role="table"
  >
    <div
      class="sc-gKclnd cFsFpL"
      role="row-group"
    >
      <div
        class="sc-iCfMLu hSbAuc"
        role="row"
      />
    </div>
    <div
      class="sc-bdvvtL hjGchb"
      role="row-group"
    />
  </div>
</div>
`;

exports[`it should render correctly with the selector function 1`] = `
<div>
  <div
    class="sc-hBUSln isUOUn"
    role="table"
  >
    <div
      class="sc-gKclnd cFsFpL"
      role="row-group"
    >
      <div
        class="sc-iCfMLu hSbAuc"
        role="row"
      >
        <div
          class="sc-jRQBWg bwoBBf"
        >
          Name
        </div>
      </div>
    </div>
    <div
      class="sc-bdvvtL hjGchb"
      role="row-group"
    >
      <div
        class="sc-ksdxgE inUDjc"
        id="row-0"
        role="row"
      >
        <div
          class="sc-gsDKAQ sc-dkPtRN sc-hKwDye iigSTN juQwsB flAfZy"
          id="cell-undefined"
          role="grid-cell"
        >
          <div>
            Joe Doe 
          </div>
        </div>
      </div>
    </div>
  </div>
  <nav
    class="sc-jrQzAO dZcCqi"
  >
    <div
      class="sc-pVTFL gUaKsY"
    >
      <select
        class="sc-furwcr biuCpP"
      >
        <option
          selected=""
          value="5"
        >
          5
        </option>
        <option
          value="10"
        >
          10
        </option>
        <option
          value="25"
        >
          25
        </option>
      </select>
      <svg
        aria-hidden="true"
        fill="currentColor"
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M5.293 7.293a1 1 0 011.414 0L10 10.586l3.293-3.293a1 1 0 111.414 1.414l-4 4a1 1 0 01-1.414 0l-4-4a1 1 0 010-1.414z"
          fill-rule="evenodd"
        />
      </svg>
    </div>
    <span
      class="sc-crHmcD sc-egiyK dOoxwS dIFsGL"
    >
      1 of 1
    </span>
    <div
      class="sc-kDTinF jFDtol"
    >
      <button
        aria-disabled="true"
        aria-label="First Page"
        class="sc-iqseJM jtEzsa"
        disabled=""
        id="pagination-first-page"
        type="button"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          style="height: 12px; width: 12px;"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M15.707 15.707a1 1 0 01-1.414 0l-5-5a1 1 0 010-1.414l5-5a1 1 0 111.414 1.414L11.414 10l4.293 4.293a1 1 0 010 1.414zm-6 0a1 1 0 01-1.414 0l-5-5a1 1 0 010-1.414l5-5a1 1 0 011.414 1.414L5.414 10l4.293 4.293a1 1 0 010 1.414z"
            fill-rule="evenodd"
          />
        </svg>
      </button>
      <button
        aria-disabled="true"
        aria-label="Previous Page"
        class="sc-iqseJM jtEzsa"
        disabled=""
        id="pagination-previous-page"
        type="button"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          style="height: 12px; width: 12px;"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M12.707 5.293a1 1 0 010 1.414L9.414 10l3.293 3.293a1 1 0 01-1.414 1.414l-4-4a1 1 0 010-1.414l4-4a1 1 0 011.414 0z"
            fill-rule="evenodd"
          />
        </svg>
      </button>
      <button
        aria-disabled="true"
        aria-label="Next Page"
        class="sc-iqseJM jtEzsa"
        disabled=""
        id="pagination-next-page"
        type="button"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          style="height: 12px; width: 12px;"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M7.293 14.707a1 1 0 010-1.414L10.586 10 7.293 6.707a1 1 0 011.414-1.414l4 4a1 1 0 010 1.414l-4 4a1 1 0 01-1.414 0z"
            fill-rule="evenodd"
          />
        </svg>
      </button>
      <button
        aria-disabled="true"
        aria-label="Last Page"
        class="sc-iqseJM jtEzsa"
        disabled=""
        id="pagination-last-page"
        type="button"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          style="height: 12px; width: 12px;"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M10.293 15.707a1 1 0 010-1.414L14.586 10l-4.293-4.293a1 1 0 111.414-1.414l5 5a1 1 0 010 1.414l-5 5a1 1 0 01-1.414 0z"
            fill-rule="evenodd"
          />
          <path
            clip-rule="evenodd"
            d="M4.293 15.707a1 1 0 010-1.414L8.586 10 4.293 5.707a1 1 0 011.414-1.414l5 5a1 1 0 010 1.414l-5 5a1 1 0 01-1.414 0z"
            fill-rule="evenodd"
          />
        </svg>
      </button>
    </div>
  </nav>
</div>
`;

exports[`it should update state when props change 1`] = `
<div>
  <div
    class="sc-hBUSln isUOUn"
    role="table"
  >
    <div
      class="sc-gKclnd cFsFpL"
      role="row-group"
    >
      <div
        class="sc-iCfMLu hSbAuc"
        role="row"
      >
        <div
          class="sc-jRQBWg bwoBBf"
        >
          Name
        </div>
      </div>
    </div>
    <div
      class="sc-bdvvtL hjGchb"
      role="row-group"
    >
      <div
        class="sc-ksdxgE inUDjc"
        id="row-0"
        role="row"
      >
        <div
          class="sc-gsDKAQ sc-dkPtRN sc-hKwDye iigSTN juQwsB flAfZy"
          id="cell-undefined"
          role="grid-cell"
        >
          <div>
            Something else
          </div>
        </div>
      </div>
    </div>
  </div>
  <nav
    class="sc-jrQzAO dZcCqi"
  >
    <div
      class="sc-pVTFL gUaKsY"
    >
      <select
        class="sc-furwcr biuCpP"
      >
        <option
          selected=""
          value="5"
        >
          5
        </option>
        <option
          value="10"
        >
          10
        </option>
        <option
          value="25"
        >
          25
        </option>
      </select>
      <svg
        aria-hidden="true"
        fill="currentColor"
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M5.293 7.293a1 1 0 011.414 0L10 10.586l3.293-3.293a1 1 0 111.414 1.414l-4 4a1 1 0 01-1.414 0l-4-4a1 1 0 010-1.414z"
          fill-rule="evenodd"
        />
      </svg>
    </div>
    <span
      class="sc-crHmcD sc-egiyK dOoxwS dIFsGL"
    >
      1 of 1
    </span>
    <div
      class="sc-kDTinF jFDtol"
    >
      <button
        aria-disabled="true"
        aria-label="First Page"
        class="sc-iqseJM jtEzsa"
        disabled=""
        id="pagination-first-page"
        type="button"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          style="height: 12px; width: 12px;"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M15.707 15.707a1 1 0 01-1.414 0l-5-5a1 1 0 010-1.414l5-5a1 1 0 111.414 1.414L11.414 10l4.293 4.293a1 1 0 010 1.414zm-6 0a1 1 0 01-1.414 0l-5-5a1 1 0 010-1.414l5-5a1 1 0 011.414 1.414L5.414 10l4.293 4.293a1 1 0 010 1.414z"
            fill-rule="evenodd"
          />
        </svg>
      </button>
      <button
        aria-disabled="true"
        aria-label="Previous Page"
        class="sc-iqseJM jtEzsa"
        disabled=""
        id="pagination-previous-page"
        type="button"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          style="height: 12px; width: 12px;"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M12.707 5.293a1 1 0 010 1.414L9.414 10l3.293 3.293a1 1 0 01-1.414 1.414l-4-4a1 1 0 010-1.414l4-4a1 1 0 011.414 0z"
            fill-rule="evenodd"
          />
        </svg>
      </button>
      <button
        aria-disabled="true"
        aria-label="Next Page"
        class="sc-iqseJM jtEzsa"
        disabled=""
        id="pagination-next-page"
        type="button"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          style="height: 12px; width: 12px;"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M7.293 14.707a1 1 0 010-1.414L10.586 10 7.293 6.707a1 1 0 011.414-1.414l4 4a1 1 0 010 1.414l-4 4a1 1 0 01-1.414 0z"
            fill-rule="evenodd"
          />
        </svg>
      </button>
      <button
        aria-disabled="true"
        aria-label="Last Page"
        class="sc-iqseJM jtEzsa"
        disabled=""
        id="pagination-last-page"
        type="button"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          style="height: 12px; width: 12px;"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M10.293 15.707a1 1 0 010-1.414L14.586 10l-4.293-4.293a1 1 0 111.414-1.414l5 5a1 1 0 010 1.414l-5 5a1 1 0 01-1.414 0z"
            fill-rule="evenodd"
          />
          <path
            clip-rule="evenodd"
            d="M4.293 15.707a1 1 0 010-1.414L8.586 10 4.293 5.707a1 1 0 011.414-1.414l5 5a1 1 0 010 1.414l-5 5a1 1 0 01-1.414 0z"
            fill-rule="evenodd"
          />
        </svg>
      </button>
    </div>
  </nav>
</div>
`;
